File: android/app-v2/src/main/java/com/smarthome/livingroom_v2/edge/PlanExecutor.kt
package com.smarthome.livingroom_v2.edge

import android.content.Context
import android.util.Log
import com.smarthome.livingroom_v2.brain.BrainClient
import com.smarthome.livingroom_v2.brain.dto.ExecutionReport
import com.smarthome.livingroom_v2.brain.dto.OnFailurePolicy
import com.smarthome.livingroom_v2.brain.dto.Plan
import com.smarthome.livingroom_v2.brain.dto.StepStatus
import com.smarthome.livingroom_v2.skill.SkillContext
import com.smarthome.livingroom_v2.skill.SkillResult

class PlanExecutor(
    private val appContext: Context,
    var edgeId: String,
    private val registry: SkillRegistry,
    private val brain: BrainClient,
) {
    data class Outcome(
        val planId: String,
        val reports: List<ExecutionReport>,
        val aborted: Boolean,
    )

    suspend fun execute(plan: Plan): Outcome {
        val reports = mutableListOf<ExecutionReport>()
        var aborted = false
        for (step in plan.steps) {
            val skill = registry.get(step.skillId)
            val result: SkillResult
            if (skill == null) {
                result = SkillResult.error("skill not registered: ${step.skillId}")
            } else {
                val ctx = SkillContext(
                    appContext = appContext,
                    edgeId = edgeId,
                    planId = plan.planId,
                    stepId = step.stepId,
                )
                result = try {
                    skill.execute(step.action, step.params, ctx)
                } catch (t: Throwable) {
                    Log.e(TAG, "skill threw id=${step.skillId} capability=${step.action}", t)
                    SkillResult.error(t.message ?: t.javaClass.simpleName)
                }
            }

            val report = ExecutionReport(
                planId = plan.planId,
                stepId = step.stepId,
                status = if (result.ok) StepStatus.OK else StepStatus.ERROR,
                message = result.message,
            )
            reports += report
            brain.report(report)

            if (!result.ok && step.onFailure == OnFailurePolicy.ABORT) {
                aborted = true
                // Mark remaining steps skipped
                for (rest in plan.steps.dropWhile { it.stepId != step.stepId }.drop(1)) {
                    val skipped = ExecutionReport(
                        planId = plan.planId,
                        stepId = rest.stepId,
                        status = StepStatus.SKIPPED,
                        message = "aborted after failure of ${step.stepId}",
                    )
                    reports += skipped
                    brain.report(skipped)
                }
                break
            }
        }
        return Outcome(planId = plan.planId, reports = reports, aborted = aborted)
    }

    companion object {
        private const val TAG = "PlanExecutor"
    }
}

File: ios/LivingRoomEdge/LivingRoomEdge/Edge/PlanExecutor.swift
import Foundation

struct PlanOutcome {
    let planId: String
    let reports: [ExecutionReport]
    let aborted: Bool
}

@MainActor
final class PlanExecutor {
    var edgeId: String
    private let registry: SkillRegistry
    private let brain: BrainClient

    init(edgeId: String, registry: SkillRegistry, brain: BrainClient) {
        self.edgeId = edgeId
        self.registry = registry
        self.brain = brain
    }

    func execute(_ plan: Plan) async -> PlanOutcome {
        var reports: [ExecutionReport] = []
        var aborted = false

        for (index, step) in plan.steps.enumerated() {
            let result: SkillResult
            if let skill = registry.get(step.skillId) {
                let ctx = SkillContext(edgeId: edgeId, planId: plan.planId, stepId: step.stepId)
                result = await skill.execute(capabilityId: step.action, params: step.params, context: ctx)
            } else {
                result = .error("skill not registered: \(step.skillId)")
            }

            let reportMessage: String?
            if let outputs = result.outputs, !outputs.isEmpty,
               let data = try? JSONSerialization.data(withJSONObject: outputs, options: [.sortedKeys]),
               let json = String(data: data, encoding: .utf8) {
                if let base = result.message, !base.isEmpty {
                    reportMessage = "\(base)\noutputs=\(json)"
                } else {
                    reportMessage = "outputs=\(json)"
                }
            } else {
                reportMessage = result.message
            }
            let report = ExecutionReport(
                planId: plan.planId,
                stepId: step.stepId,
                status: result.ok ? .ok : .error,
                message: reportMessage
            )
            reports.append(report)
            try? await brain.report(report)

            if !result.ok && step.onFailure == .abort {
                aborted = true
                for rest in plan.steps.suffix(from: index + 1) {
                    let skipped = ExecutionReport(
                        planId: plan.planId,
                        stepId: rest.stepId,
                        status: .skipped,
                        message: "aborted after failure of \(step.stepId)"
                    )
                    reports.append(skipped)
                    try? await brain.report(skipped)
                }
                break
            }
        }

        return PlanOutcome(planId: plan.planId, reports: reports, aborted: aborted)
    }
}
